tls/s2n_server_key_exchange.c

```diff
@@ -171,8 +172,142 @@ int s2n_kem_server_key_recv_parse_data(struct s2n_connection *conn, union s2n_ke
+          GUARD(s2n_stuffer_write_uint8(out, conn->secure.conn_sig_alg));
+        }
+
+        /* Add the random data to the hash */
+        GUARD(s2n_hash_init(signature_hash, conn->secure.conn_hash_alg));
+        GUARD(s2n_hash_update(signature_hash, conn->secure.client_random, S2N_TLS_RANDOM_DATA_LEN));
+        GUARD(s2n_hash_update(signature_hash, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN));
+
+        /* Add KEX specific data to the hash */
+        GUARD(s2n_hash_update(signature_hash, data_to_sign.data, data_to_sign.size));
+
+        /* prepare the digest_out blob */
+        struct s2n_blob digest_out;
+        s2n_alloc(&digest_out, S2N_MAX_DIGEST_LEN);
+
+        uint8_t digest_size;
+        GUARD(s2n_hash_digest_size(signature_hash->alg, &digest_size));
+        GUARD(s2n_hash_digest(signature_hash, digest_out.data, digest_size));
+        digest_out.size = digest_size;
+
+        conn->external_ctx.sign_status = S2N_EXTERNAL_INVOKED;
+        conn->config->external_dhe_sign(
+          (int32_t*)&(conn->external_ctx.sign_status),
+          &(conn->external_ctx.signed_hash_size),
+          &(conn->external_ctx.signed_hash),
+          (uint8_t)signature_hash->alg,
+          digest_out.data,
+          digest_out.size,
+          conn->context);
+
+        s2n_free(&digest_out);
+    }
+
+    switch(conn->external_ctx.sign_status) {
+        /* external signing has been invoked and we are waiting for the result. Return '1' to indicate that */
+        case S2N_EXTERNAL_INVOKED: {
+            return 1;
+        }
+        /* external signing has returned the result. Proceed by returning '0' */
+        case S2N_EXTERNAL_RETURNED: {
+            return 0;
+        }
+        /* error occurred with the external signing */
+        case S2N_EXTERNAL_ERROR: {
+            /* free the ctx */
+            s2n_free_external_ctx_signed_hash(conn);
+            S2N_ERROR(S2N_ERR_EXTERNAL_FAILURE);
+        }
+        /* the status is not anything we expected. Something went wrong. */
+        default: {
+            /* free the ctx */
+            s2n_free_external_ctx_signed_hash(conn);
+            S2N_ERROR(S2N_ERR_EXTERNAL_CTX_STATUS_INVALID);
+        }
+    }
+}
+
+int s2n_ecdhe_server_key_external(struct s2n_connection *conn, struct s2n_blob *data_to_sign)
+{
+    struct s2n_stuffer *out = &conn->external_ctx.ephemeral_key_io;
+    s2n_stuffer_alloc(out, 1024);
+
+    /* Generate an ephemeral key and  */
+    GUARD(s2n_ecc_generate_ephemeral_key(&conn->secure.server_ecc_params));
+
+    /* Write it out and calculate the data to sign later */
+    GUARD(s2n_ecc_write_ecc_params(&conn->secure.server_ecc_params, out, data_to_sign));
+
+    return 0;
+}
+
+int s2n_dhe_server_key_external(struct s2n_connection *conn, struct s2n_blob *data_to_sign)
+{
+    struct s2n_stuffer *out = &conn->external_ctx.ephemeral_key_io;
+    s2n_stuffer_alloc(out, 1024);
```

Comment:
as above.

what is more, if `conn->external_ctx.ephemeral_key_io` is already, allocated, we will leak the blob.  consider using `realloc.`